cachelib/allocator/CacheAllocator.h

```diff
@@ -2051,6 +2051,8 @@ auto& mmContainer = getMMContainer(tid, pid, cid);
         // TODO: only allow it for read-only items?
         // or implement mvcc
         if (candidate->markMoving(true)) {
+          // promotions should rarely fail since we already marked moving
+          mmContainer.remove(itr);
```

Comment:
As I understand promotion might fail if we are unable to allocate memory in a destination tier, right?
Should we return the candidate to mmContainer in that case? Do we have such logic?

Reply:
Correct. But currently if promotion fails we just evict the item (by nature it is a cache and data is volatile anyway). See the else statement in line 2080.

However we could do so - and I believe the logic would be as follows instead of the call to `unmarkMovingAndWakeUpWaiters`. @igchor can comment to on this.

`item.unmarkMoving();` 
`auto hdl = acquire(&item);`
`insertToMMContainer(*hdl);`
`wakeUpWaiters(item,std::move(hdl));`

Reply:
Yeah, I saw the `else` statement and just wanted to make sure I understand the current logic.
But for me it is strange logic: we are trying to promote the hot item from the second tier but incase of failure instead of just keeping it in the second tier we removes the hot item from the cache...

Reply:
Yes, that should work - I think the current logic is the result of some previous Critical Section patch implementations where. I don't see any problems with the code Daniel suggests.

Reply:
I added in the reinsert logic on promotion fail. I tested with a few configs and I believe it works. I don't see any noticeable difference in performance.